core/node/state_keeper/src/keeper.rs

```diff
@@ -938,4 +794,314 @@ impl ZkSyncStateKeeper {
+        while !is_canceled(&stop_receiver) {
+            self.process_block(&mut stop_receiver).await?;
+        }
+
+        Err(OrStopped::Stopped)
+    }
+
+    async fn process_block(
+        &mut self,
+        stop_receiver: &mut watch::Receiver<bool>,
+    ) -> Result<(), OrStopped> {
+        // If there is no open batch then start one.
+        if let Some(cursor) = self.batch_state.as_uninit() {
+            let state = self.inner.start_batch(cursor, stop_receiver).await?;
+            self.batch_state = BatchState::Init(Box::new(state));
+        }
+        let state = self.batch_state.unwrap_init_ref_mut();
+        let updates_manager = &mut state.updates_manager;
+        let batch_executor = state.batch_executor.as_mut();
+
+        if let Some(protocol_upgrade_tx) = state.protocol_upgrade_tx.take() {
+            // Protocol upgrade tx if the first tx in the block so we shouldn't do `set_l2_block_params`.
+            self.inner
+                .process_upgrade_tx(batch_executor, updates_manager, protocol_upgrade_tx)
+                .await?;
+        } else {
+            // Params for the first L2 block in the batch are pushed with batch params.
+            // For non-first L2 block they must be set manually.
+            if updates_manager.pending_executed_transactions_len() > 0 {
+                let next_l2_block_params = self
+                    .inner
+                    .wait_for_new_l2_block_params(updates_manager, stop_receiver)
+                    .await
+                    .stop_context("failed getting L2 block params")?;
+                StateKeeperInner::set_l2_block_params(updates_manager, next_l2_block_params);
+            }
+
+            // Check if batch should be sealed based on conditional seal criteria.
+            // It is not strictly required to check it here. We could try to execute the next tx and seal batch if `ExcludeAndSeal` is returned.
+            // However, this is suboptimal, checking seal criteria is much faster than executing and reverting tx.
+            if self
+                .inner
+                .should_seal_batch_conditional_criteria(updates_manager, batch_executor)
+                .await?
+            {
+                // Push the current block if it has not been done yet and this will effectively create a fictive l2 block.
+                if let Some(next_l2_block_timestamp) = updates_manager.next_l2_block_timestamp_mut()
+                {
+                    self.inner
+                        .io
+                        .update_next_l2_block_timestamp(next_l2_block_timestamp);
+                    StateKeeperInner::start_next_l2_block(updates_manager, batch_executor).await?;
+                }
+
+                self.seal_batch().await?;
+                return Ok(());
+            }
+        }
+
+        while !is_canceled(stop_receiver) {
+            let full_latency = KEEPER_METRICS.process_block_loop_iteration.start();
+            if self
```

Comment:
Bikeshedding: IMO, it'd make sense to extract iteration to a separate method. It'd allow to handle `full_latency` better (observe it in a single place), among other things.